src/main/scala/com/databricks/spark/avro/AvroRelation.scala

```diff
@@ -134,20 +129,54 @@ private[avro] class AvroRelation(
             } else {
               val firstRecord = records.next()
               val superSchema = firstRecord.getSchema // the schema of the actual record
-
               // the fields that are actually required along with their converters
-              val avroFields = superSchema.getFields.map(f => (f.name, f)).toMap
-              // the list of field names along with their conversion function
-              val fields = new ArrayBuffer[(Any => Any, String)](requiredColumns.length)
-
-              for (column <- requiredColumns) {
-                val field = avroFields.getOrElse(column,
-                  throw new SchemaConversionException(s"could not find Avro column: $column"))
-                fields.append((SchemaConverters.createConverterToSQL(field.schema), field.name))
+              val avroFieldMap = superSchema.getFields.map(f => (f.name, f)).toMap
+
+              new Iterator[Row] {
+                private[this] val baseIterator = records
+                private[this] var currentRecord = firstRecord
+                private[this] val rowBuffer = new Array[Any](requiredColumns.length)
+                // A micro optimization to avoid allocating a WrappedArray per row.
+                private[this] val bufferSeq = rowBuffer.toSeq
+
+                // An array of functions that pull a column out of an avro record and puts the
+                // converted value into the correct slot of the rowBuffer.
+                private[this] val fieldExtractors = requiredColumns.zipWithIndex.map {
+                  case (columnName, idx) =>
+                    // Spark SQL should not pass us invalid columns
+                    val field =
+                      avroFieldMap.getOrElse(
+                        columnName,
+                        throw new AssertionError(s"Invalid column $columnName"))
+                    val converter = SchemaConverters.createConverterToSQL(field.schema)
+
+                    (record: GenericRecord) => rowBuffer(idx) = converter(record.get(field.pos()))
+                }
+
+                private def advanceNextRecord() = {
```

Comment:
Just curious, but why did you choose this `advanceNext()` idiom over `def hasNext = baseIterator.hasNext` etc?